oauth2_http/java/com/google/auth/oauth2/OAuth2Credentials.java

```diff
@@ -144,75 +152,179 @@ public void getRequestMetadata(
+      refreshResult.executeIfNew(executor);
+    }
+
+    synchronized (lock) {
+      // Immediately resolve the token token if its not expired, or wait for the refresh task to
+      // complete
+      if (getState() != CacheState.Expired) {
+        return Futures.immediateFuture(value);
+      } else if (refreshResult != null) {
+        return refreshResult.task;
+      } else {
+        // Should never happen
+        return Futures.immediateFailedFuture(
+            new IllegalStateException("Credentials expired, but there is no task to refresh"));
       }
     }
   }
 
   /**
-   * Provide additional headers to return as request metadata.
+   * Atomically creates a single flight refresh token task.
    *
-   * @return additional headers
+   * <p>Only a single refresh task can be scheduled at a time. If there is an existing task, it will
+   * be returned for subsequent invocations. However if a new task is created, it is the
+   * responsibility of the caller to execute it. The task will clear the single flight slow upon
+   * completion.
    */
-  protected Map<String, List<String>> getAdditionalHeaders() {
-    return EMPTY_EXTRA_HEADERS;
+  private AsyncRefreshResult getOrCreateRefreshTask() {
+    synchronized (lock) {
+      if (refreshTask != null) {
+        return new AsyncRefreshResult(refreshTask, false);
+      }
+
+      final ListenableFutureTask<OAuthValue> task =
+          ListenableFutureTask.create(
+              new Callable<OAuthValue>() {
+                @Override
+                public OAuthValue call() throws Exception {
+                  return OAuthValue.create(refreshAccessToken(), getAdditionalHeaders());
+                }
+              });
+
+      task.addListener(
+          new Runnable() {
+            @Override
+            public void run() {
+              finishRefreshAsync(task);
+            }
+          },
+          MoreExecutors.directExecutor());
+
+      refreshTask = task;
+
+      return new AsyncRefreshResult(refreshTask, true);
+    }
   }
 
   /**
-   * Refresh these credentials only if they have expired or are expiring imminently.
+   * Async callback for committing the result from a token refresh.
    *
-   * @throws IOException during token refresh.
+   * <p>The result will be stored, listeners are invoked and the single flight slot is cleared.
    */
-  public void refreshIfExpired() throws IOException {
+  private void finishRefreshAsync(ListenableFuture<OAuthValue> finishedTask) {
     synchronized (lock) {
-      if (shouldRefresh()) {
-        refresh();
+      try {
+        this.value = finishedTask.get();
+        for (CredentialsChangedListener listener : changeListeners) {
+          listener.onChanged(this);
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      } catch (Exception e) {
+        // noop
+      } finally {
+        if (this.refreshTask == finishedTask) {
+          this.refreshTask = null;
```

Comment:
can we make refreshTask @VisibleForTesting and cover its reset to null this with some tests? To avoid some nasty regressions.

Reply:
I updated the existing tests to ensure that its managed properly. Let me know if you had something else in mind

Reply:
looks good, thx!